### ISSUE TITLE
Inference regression on to_indices

### ISSUE THREAD

**Issue by N5N3** [JuliaLang/julia#42001]
(The previous examples was deleted because of limited coverage.)

This issure is found in https://github.com/JuliaGPU/CUDA.jl/issues/1039, ~which seems to be caused by inline failure~.
I have tested it on 1.3.1, 1.4.2, 1.5.4,1.6.2, 1.7.0-dev3, master with the following code:
```julia
module test
using InteractiveUtils
macro get_infered(ex)
    esc(:((@code_typed $ex).second))
end
a = @get_infered to_indices([1], (CartesianIndex(),1,CartesianIndex(1,1)))
b = @get_infered to_indices([1], (), (CartesianIndex(),1,CartesianIndex(1,1))) 
println(b, "  --  " ,a)
a = @get_infered to_indices([1], (CartesianIndex(),1,CartesianIndex(1,1,1)))
b = @get_infered to_indices([1], (), (CartesianIndex(),1,CartesianIndex(1,1,1)))
println(b, "  --  " ,a)
a = @get_infered to_indices([1], (CartesianIndex(1),1,CartesianIndex(1,1,1)))
b = @get_infered to_indices([1], (), (CartesianIndex(1),1,CartesianIndex(1,1,1)))
println(b, "  --  " ,a)
a = @get_infered to_indices([1], (CartesianIndex(1),1,CartesianIndex(1,1,1,1)))
b = @get_infered to_indices([1], (), (CartesianIndex(1),1,CartesianIndex(1,1,1,1)))
println(b, "  --  " ,a)
a = @get_infered to_indices([1], (CartesianIndex(1),1,CartesianIndex(1,1,1), 1))
b = @get_infered to_indices([1], (), (CartesianIndex(1),1,CartesianIndex(1,1,1), 1))
println(b, "  --  " ,a)
a = @get_infered to_indices([1], (CartesianIndex(1,1),CartesianIndex(1,1,1,1)))
b = @get_infered to_indices([1], (), (CartesianIndex(1,1),CartesianIndex(1,1,1,1)))
println(b, "  --  " ,a)
a = @get_infered to_indices([1], (CartesianIndex(1,1,1),CartesianIndex(1,1,1)))
b = @get_infered to_indices([1], (), (CartesianIndex(1,1,1),CartesianIndex(1,1,1)))
println(b, "  --  " ,a)
end
```
Since the results are dependent on julia version, all them are listed:

1. on 1.3.1
```julia
Tuple{Int64,Int64,Int64}  --  Tuple{Int64,Int64,Int64}
NTuple{4,Int64}  --  NTuple{4,Int64}
NTuple{5,Int64}  --  NTuple{5,Int64}
NTuple{6,Int64}  --  NTuple{6,Int64}
NTuple{6,Int64}  --  Tuple{Int64,Int64,Int64,Vararg{Int64,N} where N} # These 2 should be identity.
NTuple{6,Int64}  --  NTuple{6,Int64}
NTuple{6,Int64}  --  NTuple{6,Int64}
```
2. on 1.4.2
```julia
Tuple{Int64,Int64,Int64}  --  Tuple{Int64,Int64,Int64}
NTuple{4,Int64}  --  Tuple{Int64,Int64,Vararg{Int64,N} where N} # These 2 should be identity.
NTuple{5,Int64}  --  Tuple{Int64,Int64,Int64,Vararg{Int64,N} where N} # These 2 should be identity.
NTuple{6,Int64}  --  Tuple{Int64,Int64,Int64,Vararg{Int64,N} where N} # These 2 should be identity.
NTuple{6,Int64}  --  Tuple{Int64,Int64,Int64,Vararg{Int64,N} where N} # These 2 should be identity.
NTuple{6,Int64}  --  Tuple{Int64,Int64,Int64,Vararg{Int64,N} where N} # These 2 should be identity.
NTuple{6,Int64}  --  Tuple{Int64,Int64,Int64,Int64,Vararg{Int64,N} where N} # These 2 should be identity.
```
3. on 1.5.4 and 1.6.2
```julia
Tuple{Int64, Int64, Int64}  --  Tuple{Int64, Int64, Int64}
NTuple{4, Int64}  --  Tuple{Int64, Int64, Vararg{Int64, N} where N} # These 2 should be identity.
NTuple{5, Int64}  --  NTuple{5, Int64}
NTuple{6, Int64}  --  Tuple{Int64, Int64, Int64, Vararg{Int64, N} where N} # These 2 should be identity.
NTuple{6, Int64}  --  Tuple{Int64, Int64, Int64, Vararg{Int64, N} where N} # These 2 should be identity.
NTuple{6, Int64}  --  NTuple{6, Int64}
NTuple{6, Int64}  --  NTuple{6, Int64}
```
4. on 1.7.0 - dev3
```julia
Tuple{Int64, Int64, Int64}  --  Tuple{Int64, Int64, Int64}
NTuple{4, Int64}  --  Tuple{Int64, Int64, Vararg{Int64}} # These 2 should be identity.
NTuple{5, Int64}  --  NTuple{5, Int64}
NTuple{6, Int64}  --  Tuple{Int64, Int64, Int64, Vararg{Int64}} # These 2 should be identity.
NTuple{6, Int64}  --  Tuple{Int64, Int64, Int64, Vararg{Int64}} # These 2 should be identity.
NTuple{6, Int64}  --  Tuple{Int64, Int64, Int64, Vararg{Int64}} # These 2 should be identity.
NTuple{6, Int64}  --  NTuple{6, Int64}
```
5. on 1.8.0-DEV.401 Commit 8364a4ccd8 (2021-08-24 22:42 UTC)
```julia
Tuple{Int64, Int64, Int64}  --  Tuple{Int64, Int64, Int64}
NTuple{4, Int64}  --  Tuple{Int64, Int64, Vararg{Int64}} # These 2 should be identity.
NTuple{5, Int64}  --  NTuple{5, Int64}
NTuple{6, Int64}  --  Tuple{Int64, Int64, Int64, Vararg{Int64}} # These 2 should be identity.
NTuple{6, Int64}  --  Tuple{Int64, Int64, Int64, Vararg{Int64}} # These 2 should be identity.
NTuple{6, Int64}  --  NTuple{6, Int64}
NTuple{6, Int64}  --  NTuple{6, Int64} 
```

To summarize, inference mismatch between `to_indices(A, args::Tuple)` and `to_indices(A, inds, args::Tuple)` exists in all tested versions.

**Comment by martinholters** [comment]
Seems to be order-dependent. In a fresh session (on master = 1.8.0-DEV.392):
```julia
julia> @code_typed Base.to_indices([1],(CartesianIndex(1), 1, CartesianIndex(1,1,1)))
CodeInfo(
1 ─ %1 = Base.getfield(I, 1, true)::CartesianIndex{1}
│ [...]
│   %7 = invoke Base.to_indices(_2::Vector{Int64}, ()::Tuple{}, %6::Tuple{Int64, Int64, CartesianIndex{3}})::Tuple{Int64, Int64, Int64, Vararg{Int64}}
└──      return %7
) => Tuple{Int64, Int64, Int64, Vararg{Int64}}
```
But in a fresh session again:
```julia
julia> code_typed(to_indices, Tuple{Vector{Int}, Tuple{}, Tuple{Int64, Int64, CartesianIndex{3}}})
1-element Vector{Any}:
 CodeInfo(
1 ─ %1 = Base.getfield(I, 1, true)::Int64
│ [...]
│   %8 = Core.tuple(%1, %2, %5, %6, %7)::NTuple{5, Int64}
└──      return %8
) => NTuple{5, Int64}

julia> @code_typed Base.to_indices([1],(CartesianIndex(1), 1, CartesianIndex(1,1,1)))
CodeInfo(
1 ─ %1  = Base.getfield(I, 1, true)::CartesianIndex{1}
│ [...]
│   %10 = Core.tuple(%5, %3, %7, %8, %9)::NTuple{5, Int64}
└──       return %10
) => NTuple{5, Int64}

**Comment by N5N3** [comment]
Well, I tried to change the order of `a` and `b` in previous code and run it in a new session.
All versions prints:
```julia
Tuple{Int64, Int64, Int64}  --  Tuple{Int64, Int64, Int64}
NTuple{4, Int64}  --  Tuple{Int64, Int64, Vararg{Int64}}  # CUDA.jl 's issue. 
NTuple{5, Int64}  --  NTuple{5, Int64}
NTuple{6, Int64}  --  NTuple{6, Int64}
NTuple{6, Int64}  --  NTuple{6, Int64}
NTuple{6, Int64}  --  NTuple{6, Int64}
NTuple{6, Int64}  --  NTuple{6, Int64}
```

**Comment by N5N3** [comment]
Here's a smaller example
```julia
module gg1
import Base: tail
NInt = NTuple{N,Int} where N
@inline gather(::Tuple{}) = ()
@inline gather(x::Tuple{Int, Vararg{Any}}) = (x[1], gather(tail(x))...)
@inline gather(x::Tuple{NInt, Vararg{Any}}) = gather((x[1]..., tail(x)...))
end
@code_typed gg.gather((1,(1,),(1,2),1))
CodeInfo(
1 ─ %1 = Base.getfield(x, 1, true)::Int64
│   %2 = Core.tuple(%1)::Tuple{Int64}
│   %3 = Core.getfield(x, 2)::Tuple{Int64}
│   %4 = Core.getfield(x, 3)::Tuple{Int64, Int64}
│   %5 = Core.getfield(x, 4)::Int64
│   %6 = Core.getfield(%3, 1)::Int64
│   %7 = Core.tuple(%6, %4, %5)::Tuple{Int64, Tuple{Int64, Int64}, Int64}
│   %8 = invoke Main.gg.gather(%7::Tuple{Int64, Tuple{Int64, Int64}, Int64})::Tuple{Int64, Int64, Vararg{Int64}}
│   %9 = Core._apply_iterate(Base.iterate, Core.tuple, %2, %8)::Tuple{Int64, Int64, Int64, Vararg{Int64}}
└──      return %9
) => Tuple{Int64, Int64, Int64, Vararg{Int64}}
```

**Comment by N5N3** [comment]
@martinholters
I tried to force `edge_matches_sv` in the following line always return `false`, and all the above examples turn to be infered precisely.
https://github.com/JuliaLang/julia/blob/6341fa5d4f7963b5b6eefe768192821e0d32cbb2/base/compiler/abstractinterpretation.jl#L424

So it seems that Julia deliberately gives up accurate Inference (to reduced inference cost?)

**Comment by martinholters** [comment]
> So it seems that Julia deliberately gives up accurate Inference (to reduced inference cost?)

Well, yes, but the alternative would be potentially inifinte inference cost. The problem is the recursive nature of `to_indices`, which calls itself with different argument types. Unless inference can be sure that this recursion terminates (on the abstract interpretation level), it bails out.
